mongo/integration/unified/logger_verification.go

```diff
@@ -0,0 +1,274 @@
+func verifyLogMessagesMatch(ctx context.Context, exp, act *logMessage) error {
+	if act == nil && exp == nil {
+		return nil
+	}
+
+	if act == nil || exp == nil {
+		return errLogDocumentMismatch
+	}
+
+	levelExp := logger.ParseLevel(exp.LevelLiteral)
+	levelAct := logger.ParseLevel(act.LevelLiteral)
+
+	// The levels of the expected log message and the actual log message
+	// must match, upto logger.Level.
+	if levelExp != levelAct {
+		return fmt.Errorf("level mismatch: want %v, got %v", levelExp, levelAct)
+	}
+
+	rawExp := documentToRawValue(exp.Data)
+	rawAct := documentToRawValue(act.Data)
+
+	// Top level data does not have to be 1-1 with the expectation, there
+	// are a number of unrequired fields that may not be present on the
+	// expected document.
+	if err := verifyValuesMatch(ctx, rawExp, rawAct, true); err != nil {
+		return fmt.Errorf("%w: %v", errLogDocumentMismatch, err)
+	}
+
+	return nil
+}
+
+// clientLogMessages is a struct representing the expected "LogMessages" for a
+// client.
+type clientLogMessages struct {
+	Client      string        `bson:"client"`
+	LogMessages []*logMessage `bson:"messages"`
+}
+
+// validateClientLogMessages will validate a single "clientLogMessages" object
+// and return an error if it is invalid, i.e. not testable.
+func validateClientLogMessages(ctx context.Context, log *clientLogMessages) error {
+	if log.Client == "" {
+		return fmt.Errorf("client is required")
+	}
+
+	if len(log.LogMessages) == 0 {
+		return fmt.Errorf("log messages are required")
+	}
+
+	for _, message := range log.LogMessages {
+		if err := validateLogMessage(ctx, message); err != nil {
+			return fmt.Errorf("message is invalid: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// validateExpectLogMessages will validate a slice of "clientLogMessages"
+// objects and return the first error encountered.
+func validateExpectLogMessages(ctx context.Context, logs []*clientLogMessages) error {
+	seenClientNames := make(map[string]struct{}) // Check for client duplication
+
+	for _, log := range logs {
+		if err := validateClientLogMessages(ctx, log); err != nil {
+			return fmt.Errorf("client is invalid: %v", err)
+		}
+
+		if _, ok := seenClientNames[log.Client]; ok {
+			return fmt.Errorf("duplicate client: %v", log.Client)
+		}
+
+		seenClientNames[log.Client] = struct{}{}
+	}
+
+	return nil
+}
+
+// logMessageValidator defines the expectation for log messages across all
+// clients.
+type logMessageValidator struct {
+	testCase *TestCase
+	err      chan error
+}
+
+// newLogMessageValidator will create a new "logMessageValidator" from a test
+// case.
+func newLogMessageValidator(testCase *TestCase) (*logMessageValidator, error) {
+	if testCase == nil {
+		return nil, fmt.Errorf("test case is required")
+	}
+
+	if testCase.entities == nil {
+		return nil, fmt.Errorf("entities are required")
+	}
+
+	validator := &logMessageValidator{
+		testCase: testCase,
+		err:      make(chan error, len(testCase.entities.clients())),
```

Comment:
Do we need to collect all errors on the `logMessageValidator`? Or, could we make this channel buffered with a size of 1 since we fail the test on the first error anyway and don't really care about the following errors?

Reply:
Even though we only care bout one error we need to buffer the errors to the size of the clients, to avoid blocking errors in the worker.